File: packages/pluto/src/clients/aws/sagemaker.ts
import { SageMakerRuntimeClient, InvokeEndpointCommand } from "@aws-sdk/client-sagemaker-runtime";
import { utils } from "@plutolang/base";
import {
  ISageMakerClient,
  SageMakerOptions,
  SageMaker as SageMakerProto,
} from "../../sagemaker.aws";
import { genAwsResourceName } from "./utils";

export class SageMaker implements ISageMakerClient {
  private readonly id: string;
  private readonly endpointName: string;

  private readonly client: SageMakerRuntimeClient;

  constructor(name: string, imageUri: string, opts?: SageMakerOptions) {
    this.id = utils.genResourceId(SageMakerProto.fqn, name);
    this.endpointName = genAwsResourceName(this.id, "endpoint");
    this.client = new SageMakerRuntimeClient({});

    imageUri;
    opts;
  }

  public async invoke(inputData: any): Promise<any> {
    const command = new InvokeEndpointCommand({
      EndpointName: this.endpointName,
      Body: JSON.stringify(inputData),
      ContentType: "application/json",
      Accept: "application/json",
    });

    try {
      const response = await this.client.send(command);
      const responseBody = new TextDecoder().decode(response.Body);
      return JSON.parse(responseBody);
    } catch (error) {
      console.error("Error invoking SageMaker endpoint", error);
      throw error;
    }
  }

  public endpointUrl(): string {
    return utils.getEnvValForProperty(this.id, "endpointUrl");
  }
}

File: packages/pluto-py/pluto_client/clients/aws/sagemaker.py
import boto3
import json
from typing import Any, Optional
from pluto_base.utils import gen_resource_id, get_env_val_for_property
from .utils import gen_aws_resource_name
from ...sagemaker import ISageMakerClient, SageMakerOptions, SageMaker as SageMakerProto


class SageMaker(ISageMakerClient):
    def __init__(
        self, name: str, image_uri: str, opts: Optional[SageMakerOptions] = None
    ):
        self.__id = gen_resource_id(SageMakerProto.fqn, name)
        self.client = boto3.client("sagemaker-runtime")

    @property
    def endpoint_name(self) -> str:
        return gen_aws_resource_name(self.__id, "endpoint")

    def invoke(self, input_data: Any) -> Any:
        response = self.client.invoke_endpoint(
            EndpointName=self.endpoint_name,
            Body=json.dumps(input_data),
            ContentType="application/json",
            Accept="application/json",
        )
        return json.loads(response["Body"].read())

    def endpoint_url(self) -> str:
        return get_env_val_for_property(self.__id, "endpointUrl")
